### PR TITLE
Ct 1670 filter by deadline

### PR DESCRIPTION
**Pull request by beccapearce**

### CODE REVIEW

File: app/views/cases/search_filters/_final_deadline.html.slim [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        = "Filter by final deadline dates"
+
+    .grid-row
+      .column-full
+        .js-quick-links
+          span#quick-link-label.bold-small
+             = "Due"
+          ul.quick-link-options
+            li.quick-link-option--first
+              - today = {day: @query.available_deadlines[:today].strftime("%d"), month: @query.available_deadlines[:today].strftime("%m"), year: @query.available_deadlines[:today].strftime("%Y")}.to_json
```
Comment (aliuk2012) [inline app/views/cases/search_filters/_final_deadline.html.slim]: Can't we set these values in the action or service? I don't think the view is the best place to have them.

File: app/views/cases/search_filters/_final_deadline.html.slim [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+              a { href="#"
+                  aria-describeby="quick-link-label"
+                  data-external-deadline-from=today
+                  data-external-deadline-to=today
+                }
+                = "Today"
+            li.quick-link-option
+              - days_3 = {day: @query.available_deadlines[:three_days].strftime("%d"), month: @query.available_deadlines[:three_days].strftime("%m"), year: @query.available_deadlines[:three_days].strftime("%Y")}.to_json
```
Comment (aliuk2012) [inline app/views/cases/search_filters/_final_deadline.html.slim]: Can't we set these values in the action or service?

File: app/views/cases/search_filters/_final_deadline.html.slim [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+              a { href="#"
+                  aria-describeby="quick-link-label"
+                  data-external-deadline-from=today
+                  data-external-deadline-to=days_3
+                }
+                = "In the next 3 days"
+            li.quick-link-option
+              - days_10 = {day: @query.available_deadlines[:ten_days].strftime("%d"), month: @query.available_deadlines[:ten_days].strftime("%m"), year: @query.available_deadlines[:ten_days].strftime("%Y")}.to_json
```
Comment (aliuk2012) [inline app/views/cases/search_filters/_final_deadline.html.slim]: Can't we set these values in the action or service?

File: app/views/cases/search_filters/_final_deadline.html.slim [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+              a { href="#"
+                  aria-describeby="quick-link-label"
+                  data-external-deadline-from=today
+                  data-external-deadline-to=days_10
+                  }
+                = "In the next 10 days"
+
+    hr
+      .form-group
```
Comment (aliuk2012) [inline app/views/cases/search_filters/_final_deadline.html.slim]: We can remove this, v3 of the date field gems now includes it

File: app/views/cases/search_filters/_final_deadline.html.slim [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        = f.gov_uk_date_field :external_deadline_from,
+                { legend_text: t('cases.search.external_deadline_from'),
+                        form_hint_text: t('cases.search.external_deadline_hint') }
+      .form-group
```
Comment (aliuk2012) [inline app/views/cases/search_filters/_final_deadline.html.slim]: We can remove this, v3 of the date field gems now includes it

**Review by aliuk2012 — CHANGES_REQUESTED** [review]

File: app/services/case_search_service.rb [hunk 6]
```diff
@@ -74,6 +76,13 @@ def remove_blank_filter_values(query_params)
                                .slice(*FILTER_ATTRIBUTES)
                                .transform_values { |values| values.grep_v '' }
                                .transform_values { |value| value.is_a?(Array) ? value.sort : value }
+                               .transform_values do |values|
+                                 if values.respond_to?(:grep_v)
+                                   values.grep_v('')
+                                 else
+                                   values
+                                 end
+                               end
```
Comment (stephenrichards) [inline app/services/case_search_service.rb]: Don't we need to take like 77 out ?

File: spec/lib/papertrail_spec.rb [hunk 7]
```diff
@@ -65,18 +65,18 @@ def type_abbreviation
       end
     end
 
-    it 'does not restore the gov_uk_dates' do
+    it 'restores the gov_uk_dates' do
```
Comment (stephenrichards) [inline spec/lib/papertrail_spec.rb]: Great!  I think we need to update the readme too on this - I think it specifically says that gov_uk dates don't work and gives a work-around.

**Review by stephenrichards — CHANGES_REQUESTED** [review]

File: app/views/cases/search_filters/_final_deadline.html.slim (hunk 4 above)
Comment (aliuk2012) [reply app/views/cases/search_filters/_final_deadline.html.slim]: We agreed to ignore this for the time being there does seem to be a bug with the govuk date gem where its displaying the fields along side each other.

**Review by stephenrichards — APPROVED** [review]

**Review by aliuk2012 — APPROVED** [review]

**Pull request merged**